### DISCUSSION TITLE
The OG AI's omniscient powers aka knowing the entire map

### DISCUSSION THREAD

**Discussion by zenseii** [ihhub/fheroes2#8767] [General]
I did some quick tests just to verify the claim that the OG AI was cheating and knows the whole map and I figured I'd share them here since I've seen some doubts about this claim.

What I found is that the AI knows the entire map on Impossible but not on Normal difficulty. It also seems like if the player uses the cheat to explore the entire map 8675309, then the AI will also see the entire map.

Both castles are empty. The AI starts with 300 or something phoenixes, the human hero starts with 1 peasant. 
This is what the map looks like:
![image](https://github.com/ihhub/fheroes2/assets/12501091/ea2c2107-4a00-4b2c-814f-30fb2197ba70)

This is impossible difficulty and the AI comes straight to the empty Blue castle on their day 5, which is probably the shortest time you need to get there. This is consistent behavior for impossible difficulty.

https://github.com/ihhub/fheroes2/assets/12501091/b2c67f67-b488-43fd-b948-22c02b4972f3

Here is the OG map file.
[AItestCheatExploration.zip](https://github.com/ihhub/fheroes2/files/15497442/AItestCheatExploration.zip)

**Comment by oleg-derevenetz** [comment]
IIRC there is such an idea that on Impossible difficulty AI has this ability since day 1, but on lower difficulties AI gains this ability later - after a certain amount of playing time (the more time the lower the difficulty).

**Reply by zenseii** [reply]
Yes, I believe I observed this too. Most likely the AI getting to see the entire map is scripted to happen after x days where x also depends on the difficulty.

Suffice to say, this post is just for curiosity's sake, there's no reason to implement such cheating in fheroes2.

**Reply by Alucard648** [reply]
AI can get this ability even earlier - if he finds, learns and casts View All spell.

**Reply by zenseii** [reply]
> AI can get this ability even earlier - if he finds, learns and casts View All spell.

The fheroes2 AI already knows how to use the View All spell to some extent in determining regions of interest at the start of a turn. I don't know about the original AI and whether it knew how to use it, and it ultimately doesn't matter.

**Comment by fheroes2bugs** [comment]
Some users of the community are agains computer cheating:
![image](https://github.com/ihhub/fheroes2/assets/145407420/397cd512-194f-4af0-992d-2bf710e0edde)

Even on "map difficulty" - "expert" and "game difficulty" - "impossible" computer should not see the whole map even on 99999999 day. No way.
